query-engine/core/src/query_graph_builder/write/utils.rs

```diff
@@ -796,33 +811,15 @@ pub fn emulate_on_update_set_null(
         QueryGraphDependency::ExecutionOrder,
     )?;
 
-    // Collect other relation fields that share at least one common foreign key with the relation key we're dealing with
-    let dependent_relation_fields: Vec<_> = dependent_model
-        .fields()
-        .relation()
-        .into_iter()
-        .filter(|rf| rf != relation_field)
-        .filter(|rf| {
-            let fks = if rf.is_inlined_on_enclosing_model() {
-                rf.scalar_fields()
-            } else {
-                rf.related_field().referenced_fields()
-            };
-
-            fks.iter().any(|fk| child_fks.contains(fk))
-        })
-        .map(|rf| match rf.is_inlined_on_enclosing_model() {
-            true => rf,
-            false => rf.related_field(),
-        })
-        .collect();
+    // Collect other relation fields that share at least one common foreign key with the relation field we're dealing with
+    let overlapping_relation_fields = collect_overlapping_relation_fields(dependent_model, relation_field);
```

Comment:
I don't understand the rationale behind extending the referential action to other relation fields: they might be pointing to a different model (bad schema design, but still possible). A referential action on a foreign key in the database would only apply to _strictly that foreign key_.

Reply:
Ah I see we check the referential actions on these fields again downstream.

Reply:
> A referential action on a foreign key in the database would only apply to strictly that foreign key.

Running those tests _without_ `relationMode = "prisma"` shows the opposite.

Consider this datamodel:

```prisma
model A {
  id   Int  @id
  b_id Int? @unique
  b    B?
}

model B {
  id   Int  @id
  a_id Int? @unique
  a    A?   @relation(fields: [a_id], references: [b_id], onDelete: SetNull)

  c C?
}

model C {
  id   Int  @id
  b_id Int? @unique
  b    B?   @relation(fields: [b_id], references: [a_id], onUpdate: Restrict)
}
```

Notice how `A`, `B` and `C` have foreign keys pointing to each other. `A.b_id` points to `B.a_id` which points to `C.b_id`.

Now say you run the following query:

```ts
await prisma.A.delete({ where { id: 1 } });
```

1. This triggers `B.a`'s `SetNull` referential action. So it sets `B.a_id` to `NULL`.
2. The issue is, `C.b` _also_ points to `B.a_id`. So we also need to maintain the referential integrity between `B` and `C`.
3. This is what `collect_overlapping_relation_fields` gathers. On this specific example, after emulating `B.a`'s referential actions, it finds other relation fields in that same model that shares the foreign key(s) that were just updated. In this case, it finds `B.c` and return `C.b` because that's were the relation is inlined.
4. So we go over model `C`, and we apply the `onUpdate` referential action of the relation field `C.b`.
5. In this case, the `C.b`'s referential action is `Restrict`. So the entire operation fails. As it does when relying on the database's referential integrity.
6. If the RA was set to `SetNull`, it would also set `C.b_id` to NULL. If it was cascade, it would also set `C.b_id` to `NULL`.

Does that make sense to you ?

Reply:
Makes sense :+1: